include/pybind11/stl_bind.h

```diff
@@ -326,6 +327,36 @@ template <typename Vector, typename Class_> auto vector_if_insertion_operator(Cl
     );
 }
 
+// Provide the buffer interface for vectors if we have data() and we have a format for it
+// GCC seems to have "void std::vector<bool>::data()" - doing SFINAE on the existence of data() is insufficient, we need to check it doesn't return void
+template <typename Vector, typename Class_>
+enable_if_t<!std::is_same<decltype(py::format_descriptor<typename Vector::value_type>::format(), std::declval<Vector>().data()), void>::value>
+vector_buffer(Class_& cl) {
+    using T = typename Vector::value_type;
+
+    try {
+        //numpy.h declares this for arbitrary types, but it may raise an exception if PYBIND11_NUMPY_DTYPE hasn't been called
+        py::format_descriptor<T>::format();
```

Comment:
Instead of checking that it *isn't* void, why don't you check that it *is* a `Vector::value_type *`?  (Ultimately that seems to align better with what actually matters in the code)